central/complianceoperator/v2/scanconfigurations/service/download.go

```diff
@@ -0,0 +1,138 @@
+	"github.com/stackrox/rox/pkg/sac/resources"
+	"github.com/stackrox/rox/pkg/zip"
+	"google.golang.org/grpc/codes"
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+func NewDownloadHandler() http.HandlerFunc {
+	snapshotStore := snapshotDS.Singleton()
+	blobStore := blobDS.Singleton()
+	handler := &downloadHandler{
+		snapshotDataStore: snapshotStore,
+		blobStore:         blobStore,
+	}
+	return handler.handle
+}
+
+type downloadHandler struct {
+	snapshotDataStore snapshotDS.DataStore
+	blobStore         blobDS.Datastore
+}
+
+func (h *downloadHandler) handle(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodGet {
+		httputil.WriteErrorf(w, http.StatusMethodNotAllowed, "Only GET requests are allowed")
+		return
+	}
+
+	id, err := parseReportID(r)
+	if err != nil {
+		httputil.WriteGRPCStyleError(w, codes.InvalidArgument, err)
+		return
+	}
+
+	ctx := r.Context()
+	slimUser := authn.UserFromContext(ctx)
+	if slimUser == nil {
+		httputil.WriteGRPCStyleError(w, codes.PermissionDenied, errors.New("Could not determine user identity from provided context"))
+		return
+	}
+
+	snapshot, found, err := h.snapshotDataStore.GetSnapshot(ctx, id)
+	if err != nil {
+		httputil.WriteGRPCStyleError(w, codes.NotFound, errors.Errorf("Error retrieving snapshot with the given id %s", id))
+		return
+	}
+	if !found {
+		httputil.WriteGRPCStyleError(w, codes.NotFound, errors.Errorf("Unable to find the snapshot with the given id %s", id))
+		return
+	}
+
+	if slimUser.GetId() != snapshot.GetUser().GetId() {
+		httputil.WriteGRPCStyleError(w, codes.PermissionDenied, errors.Errorf("Report %s cannot be downloaded by the user %s", id, slimUser.GetId()))
+		return
+	}
+
+	status := snapshot.GetReportStatus()
+	if status.GetReportNotificationMethod() != storage.ComplianceOperatorReportStatus_DOWNLOAD {
+		httputil.WriteGRPCStyleError(w, codes.InvalidArgument, errors.Errorf("Report %s did not generated a downloaded report", id))
+		return
+	}
+
+	switch status.GetRunState() {
+	case storage.ComplianceOperatorReportStatus_FAILURE:
+		httputil.WriteGRPCStyleError(w, codes.FailedPrecondition, errors.Errorf("Report %s failed", id))
+		return
+	case storage.ComplianceOperatorReportStatus_WAITING, storage.ComplianceOperatorReportStatus_PREPARING:
+		httputil.WriteGRPCStyleError(w, codes.Unavailable, errors.Errorf("Report %s is not ready for download", id))
+		return
+	}
+
+	buf := bytes.NewBuffer(nil)
+	ctx = sac.WithGlobalAccessScopeChecker(ctx,
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS),
+			sac.ResourceScopeKeys(resources.Administration),
+		),
+	)
+
+	_, exists, err := h.blobStore.Get(ctx, common.GetComplianceReportBlobPath(snapshot.GetScanConfigurationId(), id), buf)
+	if err != nil {
+		httputil.WriteGRPCStyleError(w, codes.Internal, errors.New("Failed to fetch report data"))
+		return
+	}
+	if !exists {
+		httputil.WriteGRPCStyleError(w, codes.NotFound, errors.Errorf("Report %s is not available to download", id))
```

Comment:
Maybe simply
```suggestion
		httputil.WriteGRPCStyleError(w, codes.NotFound, errors.Errorf("Report %s is unavailable", id))
```
?

Reply:
Changed to `The download of report %s is unavailable"`